src/test/java/com/yscope/logging/log4j1/TestRollingFileLogAppender.java

```diff
@@ -0,0 +1,237 @@
+    RollingFileTestAppender appender = createTestAppender(1, Integer.MAX_VALUE);
+
+    // Verify that an append-flush-append sequence triggers a rollover. We need
+    // the first append and flush to force the compressor to flush the buffered
+    // log event to the output file. The final append is to trigger the
+    // rollover.
+    int expectedNumRollovers = 0;
+    appendLogEvent(0, Level.INFO, appender);
+    assertDoesNotThrow(appender::flush);
+    validateNumSyncAndCloseEvents(appender, 0, expectedNumRollovers);
+    appendLogEvent(0, Level.INFO, appender);
+    ++expectedNumRollovers;
+    validateNumSyncAndCloseEvents(appender, 0, expectedNumRollovers);
+
+    // Verify a rollover after closing the appender
+    appender.close();
+    ++expectedNumRollovers;
+    validateNumSyncAndCloseEvents(appender, 0, expectedNumRollovers);
+  }
+
+  /**
+   * Tests the hard timeout
+   */
+  @Test
+  public void testHardTimeout () {
+    RollingFileTestAppender appender = createTestAppender(Integer.MAX_VALUE, Integer.MAX_VALUE);
+
+    // Verify no syncs occur after appending an event
+    int expectedNumSyncs = 0;
+    int expectedNumRollovers = 0;
+    int currentTimestamp = 0;
+    appendLogEvent(currentTimestamp, Level.INFO, appender);
+    validateNumSyncAndCloseEvents(appender, expectedNumSyncs, expectedNumRollovers);
+
+    // Move time forward and verify that a sync happens
+    currentTimestamp = flushInfoLevelTimeout * 60000;
+    appender.setTime(currentTimestamp);
+    ++expectedNumSyncs;
+    validateNumSyncAndCloseEvents(appender, expectedNumSyncs, expectedNumRollovers);
+
+    // Verify a sequence of INFO-ERROR events triggers a sync due to the ERROR
+    // event sooner than the timeout for the INFO event
+    appendLogEvent(currentTimestamp, Level.INFO, appender);
+    appendLogEvent(currentTimestamp, Level.ERROR, appender);
+    validateNumSyncAndCloseEvents(appender, expectedNumSyncs, expectedNumRollovers);
+    currentTimestamp += flushErrorLevelTimeout * 60000;
+    appender.setTime(currentTimestamp);
+    ++expectedNumSyncs;
+    validateNumSyncAndCloseEvents(appender, expectedNumSyncs, expectedNumRollovers);
+
+    // Verify a rollover after closing the appender
+    appender.close();
+    ++expectedNumRollovers;
+    validateNumSyncAndCloseEvents(appender, expectedNumSyncs, expectedNumRollovers);
+  }
+
+  /**
+   * Tests the soft timeout
+   */
+  @Test
+  public void testSoftTimeout () {
+    RollingFileTestAppender appender = createTestAppender(Integer.MAX_VALUE, Integer.MAX_VALUE);
+
+    // Append three events and verify syncs only happen after the timeout
+    // triggered by the last event
+    int expectedNumSyncs = 0;
+    int expectedNumRollovers = 0;
+    int currentTimestamp = 0;
+    for (int i = 0; i < 3; i++) {
+      for (int j = 0; j < 3; j++) {
+        appendLogEvent(i + j, Level.INFO, appender);
+      }
+
+      // Verify no syncs happen up to the point just before the soft timeout
```

Comment:
Can we differentiate the sync triggered by soft timeout vs hard timeout? Do we need any tests to tell the difference?

Reply:
We control which timeout to be triggered by playing around with the manual clock source, i.e., setting one to the future after a flush event occurs. In this particular case inside `testSoftTimeout` unit test, we disabled the hard timeout to ensure only soft timeout is tested. Please let us know if you feel it is important.

Reply:
SG, thanks for explaining.